subsys/power/policy/policy_residency_stm32.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2020 STMicroelectronics.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include "pm_policy.h"
+
+#include <stm32l4xx_ll_lptim.h>
+
+#define LOG_LEVEL CONFIG_SYS_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+#ifdef CONFIG_STM32_LPTIM_CLOCK_LSE
+#define LPTIM_CLOCK 32768
+#define LPTIM_TIMEBASE 0xFFFF
+#else
+#define LPTIM_CLOCK 32000
+#define LPTIM_TIMEBASE 0xF9FF
+#endif
+
+#define LPTIM_MAX_MS (((LPTIM_TIMEBASE + 1) / LPTIM_CLOCK) * 1000)
+
+#define SECS_TO_TICKS		CONFIG_SYS_CLOCK_TICKS_PER_SEC
+
+u32_t table_prescaler[8] = {
+	LL_LPTIM_PRESCALER_DIV1,
+	LL_LPTIM_PRESCALER_DIV2,
+	LL_LPTIM_PRESCALER_DIV4,
+	LL_LPTIM_PRESCALER_DIV8,
+	LL_LPTIM_PRESCALER_DIV16,
+	LL_LPTIM_PRESCALER_DIV32,
+	LL_LPTIM_PRESCALER_DIV64,
+	LL_LPTIM_PRESCALER_DIV128,
+};
+
+/* PM Policy based on STM32 SoC/Platform residency requirements */
+static const unsigned int pm_min_residency[] = {
+	CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 * SECS_TO_TICKS / MSEC_PER_SEC,
+	CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_2 * SECS_TO_TICKS / MSEC_PER_SEC,
+	CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_3 * SECS_TO_TICKS / MSEC_PER_SEC
+};
+
+enum power_states sys_pm_policy_next_state(s32_t ticks)
+{
+	int i;
+
+	if ((ticks != K_TICKS_FOREVER) && (ticks < pm_min_residency[0])) {
+		LOG_DBG("Not enough time for PM operations: %d", ticks);
+		return SYS_POWER_STATE_ACTIVE;
+	}
+
+	for (i = ARRAY_SIZE(pm_min_residency) - 1; i >= 0; i--) {
+#ifdef CONFIG_SYS_PM_STATE_LOCK
+		if (!sys_pm_ctrl_is_state_enabled((enum power_states)(i))) {
+			continue;
+		}
+#endif
+		if ((ticks == K_TICKS_FOREVER) ||
+		    (ticks >= pm_min_residency[i])) {
+			/* Set timeout for wakeup event */
+			if (ticks == K_TICKS_FOREVER) {
+				/* simply un-connect the LPTIM */
+				LL_APB1_GRP1_ForceReset(
+					LL_APB1_GRP1_PERIPH_LPTIM1);
+				LL_APB1_GRP1_DisableClock(
+					LL_APB1_GRP1_PERIPH_LPTIM1);
+			}
```

Comment:
I guess those line is the whole purpose of the STM32 specific policy. Isn't that already taken into account in the LPTIM driver? There is the same check for `K_TICKS_FOREVER` and the same action of disabling the timer.

BTW what's the corresponding use case for that? I imagine waiting for another interrupt to wake-up the CPU, for example a GPIO interrupt from a user accessible button. In that case how is the timer re-enabled? It also probably means that the time is not counted during that long sleep, but I am not sure it's an issue.

Reply:
Yes canceling the lptim here is not required, as this is done in the lptim driver, removing.

FOREVER: if the system goes to sleep without a timer to wakeup, only a reset or EXTI can resume, if the pin has been configured before entering the sleep mode.